### PR TITLE
PLT-7811 - Error logging machinery to WithTracer.hs; raise error on file write timeout during indexer close

### PR DESCRIPTION
**Pull request by willjgould**
The logging machinery we decided on as a result of https://input-output.atlassian.net/browse/PLT-7809 has been added.

Additionally, when a timeout occurs during async file writing in `FileIndexer` on shutdown, an error is raised and logged.

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense and have useful messages
    - [ ] Important changes are reflected in changelog.d of the affected packages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Targeting main unless this is a cherry-pick backport
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [ ] If relevant, reference the ADR in the PR and reference the PR in the ADR
    - [x] Reviewer requested

### CODE REVIEW

**Comment by brendanrbrown** [comment]
Noting that the merge conflicts come from the recent restructuring to migrate the "legacy" marconi-chain-index and keep only the formerly "experimental" version.

So in other words, the conflicts should be able to be resolved by just taking the change in a path `marconi-chain-index/src/Marconi/ChainIndex/Experimental/<path>` and applying it to `marconi-chain-index/src/Marconi/ChainIndex/<path>`.

File: marconi-core/src/Marconi/Core/Transformer/WithTracer.hs [hunk 1]
```diff
@@ -463,3 +477,29 @@ instance
...
+  -> IndexerError
+  -> m ()
+logIndexerError = logError IndexerFailed [SomePrism _StopIndexer]
+
+-- | A helper that allows us to add an exclusion list of error constructors
+logError
+  :: forall m e' e
+   . (MonadIO m)
+  => (e' -> e)
+  -> [SomePrism e']
```
Comment (berewt) [inline marconi-core/src/Marconi/Core/Transformer/WithTracer.hs:497]: While the variable name makes the purpose explicit, can you also add variable documentation here?

File: marconi-core/src/Marconi/Core/Transformer/WithTracer.hs [hunk 2]
```diff
@@ -463,3 +477,29 @@ instance
...
+  -> Trace m e
+  -> e'
+  -> m ()
+logError fromErr exclusionList tr err =
+  ( when (any (\(SomePrism p) -> isn't p err) exclusionList) $
```
Comment (berewt) [inline marconi-core/src/Marconi/Core/Transformer/WithTracer.hs:502]: Is it wrong? I read it as "when one error of the exclusion list isn't my error, log my error" (which will always be the case if we provide 2 different errors in the exclusion list).

```suggestion
  ( unless (any (\(SomePrism p) -> is p err) exclusionList) $
```

**Review by berewt — APPROVED** [review]

**Review by berewt — CHANGES_REQUESTED** [review]
Sorry, I approved to fast: the error filter must be fixed (unless I misinterpreted it)

File: marconi-core/src/Marconi/Core/Transformer/WithTracer.hs (hunk 2 above)
Comment (willjgould) [reply marconi-core/src/Marconi/Core/Transformer/WithTracer.hs:502]: Good catch!

**Review by berewt — APPROVED** [review]

**Pull request merged**